Index setters of HTMLSelectElement and HTMLOptionsCollection should do nothing if the requested index is too large

Ahmad-S792 opened the pull request:
#### 82369f26704dd44dc3908e100cd12d21e270807b
<pre>
Index setters of HTMLSelectElement and HTMLOptionsCollection should do nothing if the requested index is too large

Index setters of HTMLSelectElement and HTMLOptionsCollection should do nothing if the requested index is too large
<a href="https://bugs.webkit.org/show_bug.cgi?id=247797">https://bugs.webkit.org/show_bug.cgi?id=247797</a>

Reviewed by Ryosuke Niwa.

This patch is to align Webkit with Gecko / Firefox and Blink / Chromium.

Partial Merge - <a href="https://chromium.googlesource.com/chromium/blink/+/e8638d6c3d11996a2b845a31f6c255f7f17b1a89">https://chromium.googlesource.com/chromium/blink/+/e8638d6c3d11996a2b845a31f6c255f7f17b1a89</a>

This patch is to make index setter aligned with length setter. In this
patch following changes are done to &quot;setItem&quot; function:

-&gt; Do not skip processing if an option list has enough size. We don&apos;t
limit an option list length if the options were parsed by the
HTML parser.  So an option list length can be larger than
maxSelectItems.
There&apos;s no reason to skip the processing or shrink the length if
the list already has enough size.
-&gt;  Show console warnings when they do nothing due to the length limitation.

* Source/WebCore/html/HTMLSelectElement.cpp:
(HTMLSelectElement::setItem): Add logic for index setter similar to
length setter and console warning message
* LayoutTests/fast/forms/select-max-length.html: Update Test Case
* LayoutTests/fast/forms/select-max-length-expected.txt: Rebaselined

Canonical link: <a href="https://commits.webkit.org/257351@main">https://commits.webkit.org/257351@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cf0508970db6df643f9a6a97114df6ea78e41702

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98542 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7752 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31669 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107964 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168232 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102491 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8262 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85135 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91081 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/105954 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104213 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/6268 "Found 2 new test failures: css3/supports-dom-api.html, fast/text/text-edge-no-half-leading-simple.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89824 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33255 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88086 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21183 "Found 2 new test failures: css3/supports-dom-api.html, fast/text/text-edge-no-half-leading-simple.html (failure)") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/76195 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1687 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22712 "Found 4 new test failures: css3/supports-dom-api.html, fast/text/text-edge-no-half-leading-simple.html, fast/text/text-edge-no-half-leading-with-line-height-simple.html, fast/text/text-edge-with-margin-padding-border-simple.html (failure)") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1603 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/45211 "Found 3 new test failures: css3/supports-dom-api.html, fast/text/text-edge-with-margin-padding-border-simple.html, media/video-inaccurate-duration-ended.html (failure)") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5050 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6536 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42136 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2985 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/html/HTMLSelectElement.cpp:
```diff
@@ -456,8 +456,10 @@ ExceptionOr<void> HTMLSelectElement::setItem(unsigned index, HTMLOptionElement*
-    if (index > maxSelectItems - 1)
-        index = maxSelectItems - 1;
+    if (index >= length() && index >= maxSelectItems) {
+        document().addConsoleMessage(MessageSource::Other, MessageLevel::Warning, makeString("Blocked attempt to expand the option list and set an option at index. The maximum list length is ", index, maxSelectItems, '.'));
```
Why are we outputting index and maxSelectItems here.
We should probably just output maxSelectItems since we're saying "the maximum length is" prior to those two numbers.

Ahmad-S792 replied:
Sure, I will remove "index". Also any changes in Console Message?

Ahmad-S792 replied:
Updated.

rniwa reviewed (approved):
r=me

Ahmad-S792 committed:
Index setters of HTMLSelectElement and HTMLOptionsCollection should do nothing if the requested index is too large

Index setters of HTMLSelectElement and HTMLOptionsCollection should do nothing if the requested index is too large
https://bugs.webkit.org/show_bug.cgi?id=247797

Reviewed by Ryosuke Niwa.

This patch is to align Webkit with Gecko / Firefox and Blink / Chromium.

Partial Merge - https://chromium.googlesource.com/chromium/blink/+/e8638d6c3d11996a2b845a31f6c255f7f17b1a89

This patch is to make index setter aligned with length setter. In this
patch following changes are done to "setItem" function:

-> Do not skip processing if an option list has enough size. We don't
limit an option list length if the options were parsed by the
HTML parser.  So an option list length can be larger than
maxSelectItems.
There's no reason to skip the processing or shrink the length if
the list already has enough size.
->  Show console warnings when they do nothing due to the length limitation.

* Source/WebCore/html/HTMLSelectElement.cpp:
(HTMLSelectElement::setItem): Add logic for index setter similar to
length setter and console warning message
* LayoutTests/fast/forms/select-max-length.html: Update Test Case
* LayoutTests/fast/forms/select-max-length-expected.txt: Rebaselined

Canonical link: https://commits.webkit.org/257351@main
